File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/MessagesLookPresenter.java
package dev.ragnarok.fenrir.mvp.presenter;

import static dev.ragnarok.fenrir.util.Objects.nonNull;
import static dev.ragnarok.fenrir.util.Utils.getCauseIfRuntime;
import static dev.ragnarok.fenrir.util.Utils.getSelected;
import static dev.ragnarok.fenrir.util.Utils.isEmpty;
import static dev.ragnarok.fenrir.util.Utils.nonEmpty;

import android.os.Bundle;

import androidx.annotation.IntDef;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import dev.ragnarok.fenrir.domain.IMessagesRepository;
import dev.ragnarok.fenrir.domain.Repository;
import dev.ragnarok.fenrir.model.LoadMoreState;
import dev.ragnarok.fenrir.model.Message;
import dev.ragnarok.fenrir.mvp.view.IBasicMessageListView;
import dev.ragnarok.fenrir.mvp.view.IMessagesLookView;
import dev.ragnarok.fenrir.util.RxUtils;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.core.Observable;

public class MessagesLookPresenter extends AbsMessageListPresenter<IMessagesLookView> {

    private static final int COUNT = 40;
    private final IMessagesRepository messagesInteractor;
    private final int mPeerId;
    private final LOADING_STATE loadingState;
    private int mFocusMessageId;

    public MessagesLookPresenter(int accountId, int peerId, int focusTo, @Nullable Message message, @Nullable Bundle savedInstanceState) {
        super(accountId, savedInstanceState);
        messagesInteractor = Repository.INSTANCE.getMessages();
        mPeerId = peerId;
        loadingState = new LOADING_STATE((Header, Footer) -> {

            @LoadMoreState int header;
            @LoadMoreState int footer;
            switch (Header) {
                case Side.LOADING:
                    header = LoadMoreState.LOADING;
                    break;
                case Side.NO_LOADING:
                    header = LoadMoreState.CAN_LOAD_MORE;
                    break;
                default:
                    header = LoadMoreState.INVISIBLE;
                    break;
            }
            switch (Footer) {
                case Side.DISABLED:
                    footer = LoadMoreState.END_OF_LIST;
                    break;
                case Side.LOADING:
                    footer = LoadMoreState.LOADING;
                    break;
                case Side.NO_LOADING:
                    footer = LoadMoreState.CAN_LOAD_MORE;
                    break;
                default:
                    footer = LoadMoreState.INVISIBLE;
                    break;
            }
            callView(v -> v.setupHeaders(footer, header));
        });

        if (savedInstanceState == null) {
            if (nonNull(message) && focusTo == 0) {
                mFocusMessageId = 0;
                getData().clear();
                getData().add(message);

                callView(IBasicMessageListView::notifyDataChanged);
                callView(v -> v.focusTo(0));
            } else {
                mFocusMessageId = focusTo;
                initRequest();
            }
        }
    }

    @Override
    public void onGuiCreated(@NonNull IMessagesLookView viewHost) {
        super.onGuiCreated(viewHost);
        viewHost.displayMessages(getData(), lastReadId);
        loadingState.updateState();
    }

    private void initRequest() {
        int accountId = getAccountId();

        appendDisposable(messagesInteractor.getPeerMessages(accountId, mPeerId, COUNT, -COUNT / 2, mFocusMessageId, false, false)
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(this::onInitDataLoaded, this::onDataGetError));
    }

    private void onDataGetError(Throwable t) {
        loadingState.FooterDisable();
        loadingState.HeaderDisable();

        callView(v -> showError(v, getCauseIfRuntime(t)));
    }

    private void onUpDataGetError(Throwable t) {
        loadingState.FooterEnable();
        callView(v -> showError(v, getCauseIfRuntime(t)));
    }

    private void onDownDataGetError(Throwable t) {
        loadingState.HeaderEnable();
        callView(v -> showError(v, getCauseIfRuntime(t)));
    }

    public void fireDeleteForAllClick(ArrayList<Integer> ids) {
        deleteSentImpl(ids, true);
    }

    public void fireDeleteForMeClick(ArrayList<Integer> ids) {
        deleteSentImpl(ids, false);
    }

    private void deleteSentImpl(Collection<Integer> ids, boolean forAll) {
        appendDisposable(messagesInteractor.deleteMessages(getAccountId(), mPeerId, ids, forAll, false)
                .compose(RxUtils.applyCompletableIOToMainSchedulers())
                .subscribe(RxUtils.dummy(), t -> callView(v -> showError(v, t))));
    }

    public void fireFooterLoadMoreClick() {
        loadMoreUp();
    }

    public void fireHeaderLoadMoreClick() {
        loadMoreDown();
    }

    private void loadMoreDown() {
        if (!loadingState.canLoadingHeader()) return;

        Integer firstMessageId = getFirstMessageId();
        if (firstMessageId == null) {
            return;
        }

        loadingState.headerLoading();

        int accountId = getAccountId();

        int targetMessageId = firstMessageId;

        appendDisposable(messagesInteractor.getPeerMessages(accountId, mPeerId, COUNT, -COUNT, targetMessageId, false, false)
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(this::onDownDataLoaded, this::onDownDataGetError));
    }

    @Override
    protected void onActionModeDeleteClick() {
        super.onActionModeDeleteClick();
        int accountId = getAccountId();

        List<Integer> ids = Observable.fromIterable(getData())
                .filter(Message::isSelected)
                .map(Message::getId)
                .toList()
                .blockingGet();

        if (nonEmpty(ids)) {
            appendDisposable(messagesInteractor.deleteMessages(accountId, mPeerId, ids, false, false)
                    .compose(RxUtils.applyCompletableIOToMainSchedulers())
                    .subscribe(() -> onMessagesDeleteSuccessfully(ids), t -> callView(v -> showError(v, getCauseIfRuntime(t)))));
        }
    }

    @Override
    protected void onActionModeSpamClick() {
        super.onActionModeDeleteClick();
        int accountId = getAccountId();

        List<Integer> ids = Observable.fromIterable(getData())
                .filter(Message::isSelected)
                .map(Message::getId)
                .toList()
                .blockingGet();

        if (nonEmpty(ids)) {
            appendDisposable(messagesInteractor.deleteMessages(accountId, mPeerId, ids, false, true)
                    .compose(RxUtils.applyCompletableIOToMainSchedulers())
                    .subscribe(() -> onMessagesDeleteSuccessfully(ids), t -> callView(v -> showError(v, getCauseIfRuntime(t)))));
        }
    }

    private void loadMoreUp() {
        if (!loadingState.canLoadingFooter()) return;

        Integer lastMessageId = getLastMessageId();
        if (lastMessageId == null) {
            return;
        }

        loadingState.footerLoading();

        int targetLastMessageId = lastMessageId;
        int accountId = getAccountId();

        appendDisposable(messagesInteractor.getPeerMessages(accountId, mPeerId, COUNT, 0, targetLastMessageId, false, false)
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(this::onUpDataLoaded, this::onUpDataGetError));
    }

    private Integer getLastMessageId() {
        return isEmpty(getData()) ? null : getData().get(getData().size() - 1).getId();
    }

    private Integer getFirstMessageId() {
        return isEmpty(getData()) ? null : getData().get(0).getId();
    }

    @Override
    protected void onActionModeForwardClick() {
        super.onActionModeForwardClick();
        ArrayList<Message> selected = getSelected(getData());

        if (nonEmpty(selected)) {
            callView(v -> v.forwardMessages(getAccountId(), selected));
        }
    }

    @SuppressWarnings("unused")
    public void fireMessageRestoreClick(@NonNull Message message, int position) {
        int accountId = getAccountId();
        int id = message.getId();

        appendDisposable(messagesInteractor.restoreMessage(accountId, mPeerId, id)
                .compose(RxUtils.applyCompletableIOToMainSchedulers())
                .subscribe(() -> onMessageRestoredSuccessfully(id), t -> callView(v -> showError(v, getCauseIfRuntime(t)))));
    }

    private void onMessageRestoredSuccessfully(int id) {
        Message message = findById(id);

        if (nonNull(message)) {
            message.setDeleted(false);
            safeNotifyDataChanged();
        }
    }

    private void onMessagesDeleteSuccessfully(Collection<Integer> ids) {
        for (Integer id : ids) {
            Message message = findById(id);

            if (nonNull(message)) {
                message.setDeleted(true);
            }
        }

        safeNotifyDataChanged();
    }

    private void onInitDataLoaded(List<Message> messages) {
        getData().clear();
        getData().addAll(messages);
        callView(IBasicMessageListView::notifyDataChanged);
        loadingState.reset();

        int index = Utils.indexOf(messages, mFocusMessageId);
        if (index != -1) {
            callView(v -> v.focusTo(index));
        } else if (mFocusMessageId == 0) {
            callView(v -> v.focusTo(messages.size() - 1));
        }
    }

    private void onUpDataLoaded(List<Message> messages) {
        if (isEmpty(messages)) {
            loadingState.FooterDisable();
        } else {
            loadingState.FooterEnable();
        }
        int size = getData().size();

        getData().addAll(messages);
        callView(v -> v.notifyMessagesUpAdded(size, messages.size()));
    }

    private void onDownDataLoaded(List<Message> messages) {
        if (isEmpty(messages)) {
            loadingState.HeaderDisable();
        } else {
            loadingState.HeaderEnable();
        }
        getData().addAll(0, messages);
        callView(v -> v.notifyMessagesDownAdded(messages.size()));
    }

    @IntDef({Side.DISABLED, Side.NO_LOADING,
            Side.LOADING})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Side {
        int DISABLED = 0;
        int NO_LOADING = 1;
        int LOADING = 2;
    }

    private static class LOADING_STATE {
        private final NotifyChanges changes;
        private @Side
        int Header = Side.DISABLED;
        private @Side
        int Footer = Side.DISABLED;

        public LOADING_STATE(@NonNull NotifyChanges changes) {
            this.changes = changes;
        }

        public void updateState() {
            changes.updateState(Header, Footer);
        }

        public void reset() {
            Header = Side.NO_LOADING;
            Footer = Side.NO_LOADING;
            updateState();
        }

        public void footerLoading() {
            Footer = Side.LOADING;
            updateState();
        }

        public void headerLoading() {
            Header = Side.LOADING;
            updateState();
        }

        public void FooterDisable() {
            Footer = Side.DISABLED;
            updateState();
        }

        public void HeaderEnable() {
            Header = Side.NO_LOADING;
            updateState();
        }

        public void FooterEnable() {
            Footer = Side.NO_LOADING;
            updateState();
        }

        public boolean canLoadingHeader() {
            return Header == Side.NO_LOADING && Footer != Side.LOADING;
        }

        public boolean canLoadingFooter() {
            return Footer == Side.NO_LOADING && Header != Side.LOADING;
        }

        public void HeaderDisable() {
            Header = Side.DISABLED;
            updateState();
        }

        public interface NotifyChanges {
            void updateState(@Side int Header, @Side int Footer);
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/messages/messageslook/MessagesLookPresenter.kt
package dev.ragnarok.fenrir.fragment.messages.messageslook

import android.os.Bundle
import dev.ragnarok.fenrir.domain.IMessagesRepository
import dev.ragnarok.fenrir.domain.Mode
import dev.ragnarok.fenrir.domain.Repository.messages
import dev.ragnarok.fenrir.fragment.messages.AbsMessageListPresenter
import dev.ragnarok.fenrir.model.Conversation
import dev.ragnarok.fenrir.model.LoadMoreState
import dev.ragnarok.fenrir.model.Message
import dev.ragnarok.fenrir.model.Peer
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.Side
import dev.ragnarok.fenrir.util.Utils
import dev.ragnarok.fenrir.util.Utils.getCauseIfRuntime
import dev.ragnarok.fenrir.util.Utils.getSelected
import dev.ragnarok.fenrir.util.Utils.indexOf
import dev.ragnarok.fenrir.util.coroutines.CoroutinesUtils.dummy
import dev.ragnarok.fenrir.util.coroutines.CoroutinesUtils.fromIOToMain
import dev.ragnarok.fenrir.util.coroutines.CoroutinesUtils.hiddenIO

class MessagesLookPresenter(
    accountId: Long,
    peerId: Long,
    focusTo: Int,
    message: Message?,
    savedInstanceState: Bundle?
) : AbsMessageListPresenter<IMessagesLookView>(accountId, savedInstanceState) {
    private val messagesInteractor: IMessagesRepository = messages
    private val peer = Peer(peerId)
    private val loadingState: LOADING_STATE
    private var mFocusMessageId = 0

    private fun fetchConversationThenActual(needFetchChat: Boolean) {
        appendJob(
            messagesInteractor.getConversationSingle(
                accountId,
                peer.id,
                Mode.ANY
            )
                .fromIOToMain(
                    { onConversationFetched(needFetchChat, it) },
                    { onConversationFetchFail(needFetchChat, it) })
        )
    }

    private fun onConversationFetchFail(needFetchChat: Boolean, throwable: Throwable) {
        showError(view, throwable)
        view?.displayToolbarAvatar(accountId, peer)
        if (needFetchChat) {
            initRequest()
        }
    }

    private fun onConversationFetched(needFetchChat: Boolean, data: Conversation) {
        if (peer.getTitle().isNullOrEmpty()) {
            peer.setTitle(data.getDisplayTitle())
        }
        if (peer.avaUrl.isNullOrEmpty()) {
            peer.setAvaUrl(data.imageUrl)
        }
        view?.displayToolbarAvatar(accountId, peer)
        lastReadId.incoming = data.inRead
        lastReadId.outgoing = data.outRead
        if (needFetchChat) {
            initRequest()
        }
    }

    override fun onGuiCreated(viewHost: IMessagesLookView) {
        super.onGuiCreated(viewHost)
        viewHost.displayMessages(accountId, data, lastReadId)
        loadingState.updateState()
        view?.displayToolbarAvatar(accountId, peer)
    }

    override fun onMessageClick(message: Message, position: Int, x: Int?, y: Int?) {
        if (x != null && y != null) {
            resolvePopupMenu(message, position, x, y)
        }
    }

    private fun resolvePopupMenu(message: Message, position: Int, x: Int, y: Int) {
        if (Utils.isHiddenAccount(accountId) || !Settings.get()
                .main().isChat_popup_menu || Utils.countOfSelection(
                data
            ) > 0
        ) {
            return
        }
        message.isSelected = true
        safeNotifyItemChanged(position)

        view?.showPopupOptions(
            position,
            x,
            y,
            canEdit = false,
            canPin = false,
            canStar = false,
            doStar = false,
            canSpam = !message.isOut
        )
    }

    private fun initRequest() {
        appendJob(
            messagesInteractor.getPeerMessages(
                accountId = accountId,
                peerId = peer.id,
                count = COUNT,
                offset = -COUNT / 2,
                startMessageId = mFocusMessageId,
                excludeStartMessage = false,
                cacheData = false,
                rev = false
            )
                .fromIOToMain({ messages -> onInitDataLoaded(messages) }) { t ->
                    onDataGetError(
                        t
                    )
                })
    }

    private fun onDataGetError(t: Throwable) {
        loadingState.FooterDisable()
        loadingState.HeaderDisable()
        showError(getCauseIfRuntime(t))
    }

    private fun onUpDataGetError(t: Throwable) {
        loadingState.FooterEnable()
        showError(getCauseIfRuntime(t))
    }

    private fun onDownDataGetError(t: Throwable) {
        loadingState.HeaderEnable()
        showError(getCauseIfRuntime(t))
    }

    fun fireDeleteForMeClick(ids: ArrayList<Int>) {
        deleteSentImpl(ids)
    }

    private fun deleteSentImpl(ids: Collection<Int>) {
        appendJob(
            messagesInteractor.deleteMessages(
                accountId, peer.id, ids,
                forAll = false, spam = false
            )
                .fromIOToMain(dummy()) { t ->
                    showError(t)
                })
    }

    fun fireFooterLoadMoreClick() {
        loadMoreUp()
    }

    fun fireHeaderLoadMoreClick() {
        loadMoreDown()
    }

    private fun loadMoreDown() {
        if (!loadingState.canLoadingHeader()) return
        val targetMessageId = firstMessageId ?: return
        loadingState.headerLoading()
        appendJob(
            messagesInteractor.getPeerMessages(
                accountId = accountId,
                peerId = peer.id,
                count = COUNT,
                offset = -COUNT,
                startMessageId = targetMessageId,
                excludeStartMessage = true,
                cacheData = false,
                rev = false
            )
                .fromIOToMain({ onDownDataLoaded(it) }) { t ->
                    onDownDataGetError(
                        t
                    )
                })
    }

    override fun onActionModeDeleteClick() {
        super.onActionModeDeleteClick()
        val ids = ArrayList<Int>(data.size)
        for (i in data) {
            if (i.isSelected) {
                ids.add(i.getObjectId())
            }
        }
        if (ids.nonNullNoEmpty()) {
            appendJob(
                messagesInteractor.deleteMessages(
                    accountId,
                    peer.id,
                    ids,
                    forAll = false,
                    spam = false
                )
                    .fromIOToMain({ onMessagesDeleteSuccessfully(ids) }) { t ->
                        showError(getCauseIfRuntime(t))
                    })
        }
    }

    override fun onActionModeSpamClick() {
        super.onActionModeDeleteClick()
        val ids = ArrayList<Int>(data.size)
        for (i in data) {
            if (i.isSelected) {
                ids.add(i.getObjectId())
            }
        }
        if (ids.nonNullNoEmpty()) {
            appendJob(
                messagesInteractor.deleteMessages(
                    accountId, peer.id, ids,
                    forAll = false, spam = true
                )
                    .fromIOToMain({ onMessagesDeleteSuccessfully(ids) }) { t ->
                        showError(getCauseIfRuntime(t))
                    })
        }
    }

    private fun loadMoreUp() {
        if (!loadingState.canLoadingFooter()) return
        val targetLastMessageId = lastMessageId ?: return
        loadingState.footerLoading()
        appendJob(
            messagesInteractor.getPeerMessages(
                accountId = accountId,
                peerId = peer.id,
                count = COUNT,
                offset = 0,
                startMessageId = targetLastMessageId,
                excludeStartMessage = true,
                cacheData = false,
                rev = false
            )
                .fromIOToMain({ onUpDataLoaded(it) }) { t ->
                    onUpDataGetError(
                        t
                    )
                })
    }

    private val lastMessageId: Int?
        get() = if (data.isEmpty()) null else data[data.size - 1].getObjectId()
    private val firstMessageId: Int?
        get() = if (data.isEmpty()) null else data[0].getObjectId()

    override fun onActionModeForwardClick() {
        super.onActionModeForwardClick()
        val selected: ArrayList<Message> = getSelected(data)
        if (selected.nonNullNoEmpty()) {
            view?.forwardMessages(
                accountId,
                selected
            )
        }
    }

    fun fireMessageRestoreClick(message: Message) {
        val id = message.getObjectId()
        appendJob(
            messagesInteractor.restoreMessage(accountId, peer.id, id)
                .fromIOToMain({ onMessageRestoredSuccessfully(id) }) { t ->
                    showError(getCauseIfRuntime(t))
                })
    }

    private fun onMessageRestoredSuccessfully(id: Int) {
        val message = findById(id)
        if (message != null) {
            message.setDeleted(false)
            safeNotifyDataChanged()
        }
    }

    private fun onMessagesDeleteSuccessfully(ids: Collection<Int>) {
        for (id in ids) {
            findById(id)?.setDeleted(true)
        }
        safeNotifyDataChanged()
    }

    private fun onInitDataLoaded(messages: List<Message>) {
        data.clear()
        data.addAll(messages)
        view?.notifyDataChanged()
        loadingState.reset()
        val index = indexOf(messages, mFocusMessageId)
        if (index != -1) {
            view?.focusTo(index)
        } else if (mFocusMessageId == 0) {
            view?.focusTo(messages.size - 1)
        }
    }

    private fun onUpDataLoaded(messages: List<Message>) {
        if (messages.isEmpty()) {
            loadingState.FooterDisable()
        } else {
            loadingState.FooterEnable()
        }
        val size = data.size
        data.addAll(messages)
        view?.notifyMessagesUpAdded(
            size,
            messages.size
        )
    }

    private fun onDownDataLoaded(messages: List<Message>) {
        if (messages.isEmpty()) {
            loadingState.HeaderDisable()
        } else {
            loadingState.HeaderEnable()
        }
        data.addAll(0, messages)
        view?.notifyMessagesDownAdded(
            messages.size
        )
    }

    private class LOADING_STATE(private val changes: NotifyChanges) {
        @Side
        private var Header = Side.DISABLED

        @Side
        private var Footer = Side.DISABLED
        fun updateState() {
            changes.updateState(Header, Footer)
        }

        fun reset() {
            Header = Side.NO_LOADING
            Footer = Side.NO_LOADING
            updateState()
        }

        fun footerLoading() {
            Footer = Side.LOADING
            updateState()
        }

        fun headerLoading() {
            Header = Side.LOADING
            updateState()
        }

        fun FooterDisable() {
            Footer = Side.DISABLED
            updateState()
        }

        fun HeaderEnable() {
            Header = Side.NO_LOADING
            updateState()
        }

        fun FooterEnable() {
            Footer = Side.NO_LOADING
            updateState()
        }

        fun canLoadingHeader(): Boolean {
            return Header == Side.NO_LOADING && Footer != Side.LOADING
        }

        fun canLoadingFooter(): Boolean {
            return Footer == Side.NO_LOADING && Header != Side.LOADING
        }

        fun HeaderDisable() {
            Header = Side.DISABLED
            updateState()
        }

        interface NotifyChanges {
            fun updateState(@Side Header: Int, @Side Footer: Int)
        }
    }

    fun fireTranscript(voiceMessageId: String?, messageId: Int) {
        appendJob(
            messages.recogniseAudioMessage(accountId, messageId, voiceMessageId)
                .hiddenIO()
        )
    }

    companion object {
        private const val COUNT = 40
    }

    init {
        loadingState = LOADING_STATE(object : LOADING_STATE.NotifyChanges {
            override fun updateState(Header: Int, Footer: Int) {
                @LoadMoreState val header: Int = when (Header) {
                    Side.LOADING -> LoadMoreState.LOADING
                    Side.NO_LOADING -> LoadMoreState.CAN_LOAD_MORE
                    else -> LoadMoreState.INVISIBLE
                }
                @LoadMoreState val footer: Int = when (Footer) {
                    Side.DISABLED -> LoadMoreState.END_OF_LIST
                    Side.LOADING -> LoadMoreState.LOADING
                    Side.NO_LOADING -> LoadMoreState.CAN_LOAD_MORE
                    else -> LoadMoreState.INVISIBLE
                }
                view?.setupHeaders(
                    footer,
                    header
                )
            }

        })
        if (savedInstanceState == null) {
            if (message != null && focusTo == 0) {
                mFocusMessageId = 0
                data.clear()
                data.add(message)
                view?.notifyDataChanged()
                view?.focusTo(0)
                fetchConversationThenActual(false)
            } else {
                mFocusMessageId = focusTo
                fetchConversationThenActual(true)
            }
        }
    }
}